src/hardhat.ts

```diff
@@ -0,0 +1,68 @@
+import { CHAINS } from './generated/chains';
+import { toUpperSnakeCase } from './utils/strings';
+import { Chain, HardhatEtherscanConfig, HardhatNetworksConfig } from './types';
+
+export function buildEnvVariables(): string[] {
+  return CHAINS
+    .filter((chain) => chain.explorer?.api?.key?.required)
+    .map((chain) => buildEtherscanApiKeyName(chain));
+}
+
+export function buildEtherscanApiKeyName(chain: Chain): string {
+  return `${toUpperSnakeCase(chain.alias)}_ETHERSCAN_API_KEY`;
+}
+
+// https://hardhat.org/hardhat-runner/plugins/nomicfoundation-hardhat-verify#multiple-api-keys-and-alternative-block-explorers
+export function buildEtherscanConfig(): HardhatEtherscanConfig {
+  // Not usable outside of a Node.js environment
+  if (typeof window !== 'undefined') {
+    return { apiKey: {}, customChains: [] };
+  }
+
+  return CHAINS.reduce((etherscan, chain) => {
+    if (!chain.explorer || !chain.explorer.api) {
+      return etherscan;
+    }
+
+    const explorer = chain.explorer;
+    const apiKey = chain.explorer.api.key;
+
+    const apiKeyEnvName = buildEtherscanApiKeyName(chain);
+    const apiKeyValue = apiKey.required ? chain.alias : process.env[apiKeyEnvName];
+
+    if (apiKey.hardhatEtherscanAlias) {
+      etherscan.apiKey[apiKey.hardhatEtherscanAlias] = apiKeyValue || '';
+      return etherscan;
+    }
+
+    etherscan.customChains.push({
+      network: chain.alias,
+      chainId: Number(chain.id),
+      urls: {
+        apiURL: explorer.api?.url || '',
```

Comment:
The point of creating a custom chain for hardhat-etherscan is to verify the deployments using the API, so this shouldn't be allowed to be undefined

Reply:
Good catch. `chain.explorer.api` being falsey was checked at the start 👍🏻  (I've dropped the `?` and `|| ''`)